resources/type-system/inference.md

```diff
@@ -970,167 +970,232 @@ with respect to `L` under constraints `C0`
-      succeeded. Expression soundness follows trivially because the completed
-      value is an instance of `T_2`, and the static type is `T_2`._
+_The coercion operation satisfies the soundness invariant that the static type
+of `m_2` is guaranteed to be a subtype of `T`. A proof of this is sketched
+below._
+
+_Coercions are used in most situations where the existing spec calls for an
+assignability check._
 
-    - Otherwise, `m_2` completes with an exception. _In other words, the
-      implicit cast from `dynamic` to `T_2` has failed. This trivially satisfies
-      expression soundess, since expression soundness is only concerned with the
-      value of the expression when it completes normally._
+Coercion of an elaborated expression `m_1` to type `T` produces `m_2`, which is
+determined as follows:
 
-  - Otherwise, if `T_1` is an interface type that contains a method called
-    `call` with type `U`, and `U <: T_2`, then let `m_2` be an expression
-    artifact with static type `T_2`, whose runtime behavior is as follows:
+- Let `T_1` be the static type of `m_1`.
 
-    - Execute expression artifact `m_1`, and let `o_1` be the resulting
-      value. _By expression soundness, `o_1` will be an instance of the type
-      `T_1`._
+- If `T_1 <: T`, then let `m_2` be `m_1`. _Since `m_1` and `m_2` are the same,
+  the static type of `m_2` is `T_1`. Therefore, since `T_1 <: T`, the soundness
+  invariant is satisfied._
 
-    - Let `o_2` be the result of tearing off the `call` method of `o_1`. _This
-      is guaranteed to succeed since `o_1` is an instance of `T_1`, and it is
-      guaranteed by tear-off soundness that `o_2` will be an instance of `U`._
+- Otherwise, if `T_1` is `dynamic`, then let `m_2` be
+  `@IMPLICIT_CAST<T>(m_1)`. _Since `@IMPLICIT_CAST<T>(m_1)` has a static type of
+  `T`, the soundness invariant is satisfied._
 
-    - `m_2` completes with the value `o_2`. _Expression soundness follows from
-      the fact that `o_2` is an instance of `U` and `U <: T_2`._
+- Otherwise, if `T_1` is an interface type that contains a method called `call`
+  with type `U`, and `U <: T`, then let `m_2` be `m_1.call`. _Since `m_1.call`
+  has static type `U`, and `U <: T`, the soundness invariant is satisfied._
 
-  - _TODO(paulberry): add more cases to handle implicit instantiation of generic
-    function types, and `call` tearoff with implicit instantiation._
+- _TODO(paulberry): add more cases to handle implicit instantiation of generic
+  function types, and `call` tearoff with implicit instantiation._
 
-  - Otherwise, there is a compile-time error. _We have an expression artifact of
-    type `T_1` in a situation that requires `T_2`, which isn't a supertype, nor
-    is there a coercion available, so it's a type error._
+- Otherwise, there is a compile-time error. _We have an expression of type `T_1`
+  in a situation that requires `T`, which isn't a supertype, nor is there a
+  coercion available, so it's a type error._
```

Comment:
Yes, this should work ok. I haven’t written the spec text for it yet, but what we do in the case of an assignment to a promoted local variable is:
- Infer the RHS of the assignment using the promoted type of the local variable as the context.
- Coerce the resulting elaborated expression to the (unpromoted) static type of the local variable.
- If the static type of the coerced elaborated expression is not a subtype of the variable’s promoted type, the variable is demoted.

So in your second example, where we promote a local variable from `Function` to `int Function(int)`, and then assign a callable object with a `call` method with type `String Function(String)`, the target type for the coercion is `Function`, so the call tearoff happens, and the coerced expression has static type `String Function(String)`, and the variable gets demoted.

Reply:
Super. That also means that my idea of just putting a "try to coerce" step in between any two inference steps won't work, because the context type for inference isn't always the same as the context type for coercion.
(But I guess we could pass both to the coercion step, which then recurses with one on the inference, and then coerces the result to the other.)